### ISSUE TITLE
AccessViolation 0xC0000005 was thrown while calling AddHostObjectToScript

### ISSUE THREAD

**Issue by hez2010** [microsoft/microsoft-ui-xaml#4377]
**Describe the bug**

**Steps to reproduce the bug**

1. Create a WinUI 3 project (either UWP or Desktop)
2. Follow https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.core.corewebview2.addhostobjecttoscript
3. `AccessViolation ` was thrown while calling `AddHostObjectToScript`

**Expected behavior**

No exception was thrown.

**Version Info**

NuGet package version: WASDK 1.0, WinUI 3

Windows app type:
| UWP              | Win32            |
| :--------------- | :--------------- |
| Yes | Yes |

| Windows 10 version                  | Saw the problem? |
| :--------------------------------- | :-------------------- |
| Windows 11 (22000) | Yes |
| Insider Build (21322)              | Yes   |
| October 2020 Update (19042)        |    |
| May 2020 Update (19041)            |    |
| November 2019 Update (18363)       |    |
| May 2019 Update (18362)            |    |
| October 2018 Update (17763)        |    |
| April 2018 Update (17134)          |    |
| Fall Creators Update (16299)       |    |
| Creators Update (15063)            |    |

| Device form factor | Saw the problem? |
| :----------------- | :--------------- |
| Desktop            | Yes |
| Xbox               |  |
| Surface Hub        |  |
| IoT                |  |

**Comment by StephenLPeters** [comment]
@adambarlow  and @codendone

**Comment by codendone** [comment]
This API is not yet implemented in the core WebView2 WinRT API.

**Comment by michael-hawker** [comment]
@codendone is this scheduled on the roadmap yet for an upcoming release? It's already working for WPF, right?

This was a pain to work-around, and I've been hesitant to make any sweeping changes to the Monaco component as it's a significant fork with our workaround vs. having basically just a method rename between UWP and WinUI 3.

**Comment by codendone** [comment]
@michael-hawker Yes, the underlying functionality has existed for a while. The issue holding it back from WinUI is not having support for this API via the WinRT wrappers.

This is now fixed internally and will be part of the next release. It will require a little indirection, since the underlying functionality currently still only supports IDispatch objects, rather than IInspectable objects which WinRT APIs required. So use requires going via a COM interface, which will look something like this:
```
  var webview2Interop = (ICoreWebView2Interop)myWebView2.CoreWebView2;
  webview2Interop.AddHostObjectToScript("name", dispatchObject);
```

The expectation is eventually the WinRT API will be supported (as long as the app passes in an IInspectable), but the interop interface shown above unblocks the scenario.

**Comment by azchohfi** [comment]
This fix made it to Reunion 0.5.0-prerelease, but it will be out for 0.5.0.

**Comment by mjfusa** [comment]
Can still repro in Reunion 0.5.6. Anyone have this working? Tried workaround, but can't find ICoreWebView2Interop. @codendone where does ICoreWebView2Interop come from? WebView2 Runtime 90.0.818.56.

**Comment by michael-hawker** [comment]
Yeah, we're still waiting on this so we can more easily [port the Monaco wrapper over to WinUI 3](https://github.com/hawkerm/monaco-editor-uwp/pull/32) without a big work-around.

**Comment by dotMorten** [comment]
Any update to this issue now that 0.8.0 is out?

**Comment by michael-hawker** [comment]
@azchohfi did you bother testing this out with the Toolkit sample app or still using the workaround? @stammen know you've been tracking some other WebView2 stuff too, not sure if this is on your radar as well or not.

**Comment by azchohfi** [comment]
@codendone any updates on this? Even using the ICoreWebView2Interop doesn't unblock the scenario. The method itself returns, but then calling the JS code to get the object and call a method on it doesn't work at all. Example:
```js
const Debug = chrome.webview.hostObjects.Debug;
await Debug.log("ButtonClick");
```
```
Uncaught (in promise) Error: Unable to call method on non-function.
    at <anonymous>:1:17880

Stack:
(anonymous) | @ | VM65:1
Promise.then (async) |   |  
createPublicObject | @ | VM65:1
applyHostFunction | @ | VM65:1
apply | @ | VM65:1
myFunction | @ | index.html:32
```


```
Uncaught (in promise) TypeError: Cannot read property 'has_object' of undefined
    at <anonymous>:1:19434

Stack:
(anonymous) | @ | VM65:1
Promise.then (async) |   |  
_then | @ | VM65:1
```

**Comment by hez2010** [comment]
I've upgrade to 2.8.0-prerelease, but this time it produces AV instead of NotImplementedException.

**Comment by hez2010** [comment]
I also tested against WinUI 3 Desktop app and it also happens to WinUI 3 Desktop Apps.

**Comment by rsegner** [comment]
I've tested this using WinUI3 desktop WindowsAppSDK v1 preview 3 and still getting the access violation error

**Comment by krschau** [comment]
There are a few separate issues happening here:
1. The instances where you're seeing an access violation *should* be giving a NotImplementedException. That is being tracked in the [WebView2Feedback repo](https://github.com/MicrosoftEdge/WebView2Feedback/issues/2144).
2. To call the AddHostObjectToScript API, you *do* need to call it via the [ICoreWebView2Interop](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/winrt/interop/icorewebview2interop?view=webview2-winrt-1.0.1072.54) interface. Try something like this:
``` cs
// Instead of
// MyWebView.CoreWebView2.AddHostObjectToScript("bridge", new Bridge());
var core = MyWebView.CoreWebView2;
var interop = (ICoreWebView2Interop)(object)core;
interop.AddHostObjectToScript("bridge", new Bridge());
```
3. This will fail in WinUI 2 because .Net UWP does not support Object -> VARIANT marshalling. You will probably see this error: "`System.PlatformNotSupportedException: Marshalling between VARIANT and System.Object is not supported`". We will track this internally, but if someone needs this in WinUI 2 please feel free to open an issue for it.

As long as ICoreWebView2Interop is used, this should work in WinUI 3. If it does not, please open a new issue and we can look further.

Since the original issue was addressed (adding the interop allows a WinUI 3 developer to call the API) and the additional issues are being tracked separately, I'm going to close this issue.

**Comment by michael-hawker** [comment]
@krschau this issue was filed before the UWP version existed. We also need this on WinUI 2, please leave this issue open if it's not working there still as well.

The goal is to be able to target abstractions on top of WebView2 across WinForms, WFP, UWP, and WinUI 3.

**Comment by YunyiZhang** [comment]
> There are a few separate issues happening here:
> 
> 1. The instances where you're seeing an access violation _should_ be giving a NotImplementedException. That is being tracked in the [WebView2Feedback repo](https://github.com/MicrosoftEdge/WebView2Feedback/issues/2144).
> 2. To call the AddHostObjectToScript API, you _do_ need to call it via the [ICoreWebView2Interop](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/winrt/interop/icorewebview2interop?view=webview2-winrt-1.0.1072.54) interface. Try something like this:
> 
> ```cs
> // Instead of
> // MyWebView.CoreWebView2.AddHostObjectToScript("bridge", new Bridge());
> var core = MyWebView.CoreWebView2;
> var interop = (ICoreWebView2Interop)(object)core;
> interop.AddHostObjectToScript("bridge", new Bridge());
> ```
> 
> 3. This will fail in WinUI 2 because .Net UWP does not support Object -> VARIANT marshalling. You will probably see this error: "`System.PlatformNotSupportedException: Marshalling between VARIANT and System.Object is not supported`". We will track this internally, but if someone needs this in WinUI 2 please feel free to open an issue for it.
> 
> As long as ICoreWebView2Interop is used, this should work in WinUI 3. If it does not, please open a new issue and we can look further.
> 
> Since the original issue was addressed (adding the interop allows a WinUI 3 developer to call the API) and the additional issues are being tracked separately, I'm going to close this issue.
How to call ICoreWebView2Interop,  I used the following definition of [ICoreWebView2Interop](https://github.com/microsoft/microsoft-ui-xaml/blob/main/dev/WebView2/TestUI/WebView2BasicPage.xaml.cs#L1955), but it still didn't work

**Comment by YunyiZhang** [comment]
How to call [ICoreWebView2Interop](https://github.com/microsoft/microsoft-ui-xaml/blob/main/dev/WebView2/TestUI/WebView2BasicPage.xaml.cs#L1955), I used the following definition of ICoreWebView2Interop, but it still didn't work
